### PR TITLE
SQL-1067: Refactor Mongohandle to use fine grain locking

### PR DESCRIPTION
**Pull request by pmeredit**
(also add Option<&MongoConnection> to next())

The impetus for this was actually that I needed to add MongoConnection to next to make the fields call (SQLColumns) not completely terrible, this way we can lazily get the schema info.

The other benefit is this will reduce lock contention. I don't think anything here will cause a deadlock, try to reason about that. But I think our biggest double lock holds are read on attributes with simultaneous write to errors. Nothing is trying to write to attributes and errors at the same time, so I think we should be good. 

Actually, most lock grabs now only exist for the span of one statement. But there are still a few to look at.

### CODE REVIEW

File: core/src/util/decimal128.rs [hunk 1]
```diff
@@ -133,7 +133,7 @@ impl Decimal128Plus for [u8; 16] {
         };
```
Comment (pmeredit) [inline core/src/util/decimal128.rs:133]: clippy on nightly flagged these.

File: core/src/util/mod.rs [hunk 2]
```diff
@@ -1,6 +1,15 @@
 mod decimal128;
+use bson::{doc, Document};
 pub use decimal128::Decimal128Plus;
 
+// Replaces SQL wildcard characters with associated regex
```
Comment (pmeredit) [inline core/src/util/mod.rs:5]: This was moved because it's also needed in fields.rs, though the main implementation will be in another patch, I don't feel like undoing this :P

File: odbc/src/api/col_attr_describe_tests.rs [hunk 3]
```diff
@@ -15,10 +15,10 @@ mod unit {
     // has been called).
```
Comment (pmeredit) [inline odbc/src/api/col_attr_describe_tests.rs:15]: All test files should be safely ignorable, the only change is to creating handles (and some memory cleanup).

**Review by mattChiaravalloti — APPROVED** [review]
Looks great! 🔥 It's a big refactor but the changes all look good. I left a suggestion for going even further with the refactor to extract a common pattern but that is totally optional. Agreed with your deadlock assessment.

**Review by jchemburkar — APPROVED** [review]
lgtm!

File: odbc/src/api/data.rs [hunk 4]
```diff
@@ -664,16 +688,21 @@ pub unsafe fn format_bson_data(
...
                 return SqlReturn::SUCCESS_WITH_INFO;
             }
             *str_len_or_ind_ptr = odbc_sys::NULL_DATA;
-            indices.insert(col_num, CachedData::Fixed);
+            stmt.var_data_cache
+                .write()
+                .unwrap()
+                .as_mut()
+                .unwrap()
+                .insert(col_num, CachedData::Fixed);
```
[the remark this answers is not in the document]
Comment (pmeredit) [reply odbc/src/api/data.rs:705]: I could add like a var_data_cache_insert method, that would be good!

**Comment by pmeredit** [comment]
evergreen merge
